### PR TITLE
fix(application-system): using aws-sdkv3 instead of v2 and refactoring

### PR DESCRIPTION
**Pull request by HjorturJ**
# ...

Attach a link to issue if relevant

## What

Refactor usages of s3 clients into a single service class within the template-api-modules backend.

## Why

To consolidate s3 useage to make it easier to pinpoint bugs in the future and hopefully make it easier to make new stuff with predefined functions for common use cases of s3 buckets.

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
	- Updated the header's background color for improved visibility.
	- Introduced new methods in the attachment service for better handling of file content and URLs.
	- Added a new AWS service to streamline interactions with Amazon S3.
	- Enhanced various services to utilize the new AWS service for improved file operations.
	- Comprehensive unit tests for AWS service functionalities, including file retrieval and presigned URL generation.

- **Bug Fixes**
	- Improved error handling and file content processing in several services.

- **Chores**
	- Updated dependencies in `package.json` to enhance AWS functionality.

### CODE REVIEW

File: libs/application/template-api-modules/src/lib/modules/templates/health-insurance/health-insurance.service.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { Injectable } from '@nestjs/common'
+import { Inject, Injectable } from '@nestjs/common'
```
Comment (jonnigs) [inline libs/application/template-api-modules/src/lib/modules/templates/health-insurance/health-insurance.service.ts:1]: Is this import used?

File: libs/application/template-api-modules/src/lib/modules/templates/health-insurance-declaration/attachments/provider.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { ApplicationWithAttachments as Application } from '@island.is/application/types'
 import { AttachmentS3Service } from '../../../shared/services'
-import { Injectable } from '@nestjs/common'
+import { Inject, Injectable } from '@nestjs/common'
```
Comment (jonnigs) [inline libs/application/template-api-modules/src/lib/modules/templates/health-insurance-declaration/attachments/provider.ts:3]: Unused import

File: libs/application/template-api-modules/src/lib/modules/templates/data-protection-complaint/attachments/providers/applicationAttachmentProvider.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
 import { ApplicationWithAttachments as Application } from '@island.is/application/types'
 import { DocumentInfo } from '@island.is/clients/data-protection-complaint'
 import { AttachmentType } from '../../models/attachments'
-import { AttachmentS3Service } from '../attachment-s3.service'
-import { Injectable } from '@nestjs/common'
+import { Inject, Injectable } from '@nestjs/common'
```
Comment (jonnigs) [inline libs/application/template-api-modules/src/lib/modules/templates/data-protection-complaint/attachments/providers/applicationAttachmentProvider.ts:4]: Inject unused?

File: libs/application/template-api-modules/src/lib/modules/templates/complaints-to-althingi-ombudsman/attachments/providers/applicationAttachmentProvider.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import { ApplicationWithAttachments as Application } from '@island.is/applicatio
 import { DocumentInfo } from '@island.is/clients/althingi-ombudsman'
 import { AttachmentType } from '../../models/attachments'
 import { AttachmentS3Service } from '../../../../shared/services'
-import { Injectable } from '@nestjs/common'
+import { Inject, Injectable } from '@nestjs/common'
```
Comment (jonnigs) [inline libs/application/template-api-modules/src/lib/modules/templates/complaints-to-althingi-ombudsman/attachments/providers/applicationAttachmentProvider.ts:5]: Inject unused?

File: libs/application/template-api-modules/src/lib/modules/templates/children-residence-change/children-residence-change.service.ts [hunk 5]
```diff
@@ -13,7 +13,6 @@ import {
 } from '@island.is/application/templates/family-matters-core/utils'
```
Comment (jonnigs) [inline libs/application/template-api-modules/src/lib/modules/templates/children-residence-change/children-residence-change.service.ts:13]: Ég er með aðra PR sem eyðir þessari skrá í burtu, sem mun conflict-a við þessa. Þessi umsókn er ekki í notkun (er verið að nota children-residence-change-v2), er hægt að skilja þetta eftir óbreytt?

File: libs/application/template-api-modules/src/lib/modules/shared/services/attachment-s3.service.ts [hunk 6]
```diff
@@ -43,45 +51,94 @@ export class AttachmentS3Service {
...
   ): Promise<AttachmentData[]> {
     return await Promise.all(
       answers.map(async ({ key, name }) => {
-        const url = (
-          application.attachments as {
-            [key: string]: string
-          }
-        )[key]
-
+        const url = this.getFilenameFromApplication(application, key)
```
Comment (jonnigs) [inline libs/application/template-api-modules/src/lib/modules/shared/services/attachment-s3.service.ts:58]: Tiny bit confusing that `getFilename...` returns a url

File: libs/application/template-api-modules/src/lib/modules/templates/health-insurance/health-insurance.service.ts (hunk 1 above)
Comment (HjorturJ) [reply libs/application/template-api-modules/src/lib/modules/templates/health-insurance/health-insurance.service.ts]: Nope, removed.

File: libs/application/template-api-modules/src/lib/modules/templates/health-insurance-declaration/attachments/provider.ts (hunk 2 above)
Comment (HjorturJ) [reply libs/application/template-api-modules/src/lib/modules/templates/health-insurance-declaration/attachments/provider.ts:3]: Removed, takk 😄

File: libs/application/template-api-modules/src/lib/modules/templates/data-protection-complaint/attachments/providers/applicationAttachmentProvider.ts (hunk 3 above)
Comment (HjorturJ) [reply libs/application/template-api-modules/src/lib/modules/templates/data-protection-complaint/attachments/providers/applicationAttachmentProvider.ts]: Removed

File: libs/application/template-api-modules/src/lib/modules/templates/children-residence-change/children-residence-change.service.ts (hunk 5 above)
Comment (HjorturJ) [reply libs/application/template-api-modules/src/lib/modules/templates/children-residence-change/children-residence-change.service.ts]: Sure, reverted.

File: libs/application/template-api-modules/src/lib/modules/shared/services/attachment-s3.service.ts (hunk 6 above)
Comment (HjorturJ) [reply libs/application/template-api-modules/src/lib/modules/shared/services/attachment-s3.service.ts]: Satt, breytti variable name í eh skiljanlegra vonandi, copy-paste function frá öðrum class 😅

File: libs/nest/aws/src/lib/aws.service.ts [hunk 7]
```diff
@@ -1,74 +1,173 @@
...
+    @Inject(LOGGER_PROVIDER) protected readonly logger: Logger,
+  ) {}
 
-  constructor() {
-    this.s3 = new S3()
+  public async getFile(
+    BucketKeyPairOrFilename: BucketKeyPair | string,
+  ): Promise<GetObjectCommandOutput | undefined> {
+    const { bucket, key } = this.getBucketKey(BucketKeyPairOrFilename)
+    return await this.getFileResponse(bucket, key)
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.ts]: Ah, thanks, good to know. Fixed 😄

File: libs/nest/aws/src/lib/aws.service.ts [hunk 8]
```diff
@@ -1,74 +1,173 @@
...
+
+      const { Location: url } = await parallelUploads3.done()
+
+      if (!url)
+        throw new Error('No location url found after uploading file to S3')
+
+      return url
+    } catch (error) {
+      this.logger.error('Error occurred while uploading file to S3')
+      this.logger.error(error)
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.ts]: ❤️ Fixed.

File: package.json [hunk 9]
```diff
@@ -69,8 +69,11 @@
     "@apollo/gateway": "2.1.1",
     "@apollo/subgraph": "2.1.1",
     "@apollo/utils.keyvadapter": "3.0.0",
+    "@aws-sdk/client-s3": "3.637.0",
```
Comment (saevarma) [inline package.json:72]: I'm wondering about `@aws-sdk/...` versions if we should update `@aws-sdk/client-sqs` and `@aws-sdk/credential-provider-node` so all `@aws-sdk` versions is matching.

Comment (HjorturJ) [reply package.json:72]: I've upgraded the credentials provider, but the client-sqs has multiple breaking changes between 3.1 and 3.6.
I fixed a few simple issues before hitting some more complex ones that would require more investigation and probably refactoring 🤔

Comment (HjorturJ) [reply package.json:72]: Ok looked into it some more and think I managed to solve all the issues with the sqs upgrade, feel pretty confident after I noticed that the only place the issue was was just in a log statement 😅

**Pull request closed without merging**